### PR TITLE
Update with Limit Plan

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

This PR adds support for Limit on an Update statement on sharded keyspace.

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/engine/dml_with_input.go [hunk 1]
```diff
@@ -26,12 +26,12 @@ import (
...
+var _ Primitive = (*DMLWithInput)(nil)
 
-const DmVals = "dm_vals"
+const DmlVals = "dml_vals"
 
-// DeleteWithInput represents the instructions to perform a delete operation based on the input result.
-type DeleteWithInput struct {
+// DMLWithInput represents the instructions to perform a DML operation based on the input result.
+type DMLWithInput struct {
 	Delete Primitive
```
Comment (systay) [inline go/vt/vtgate/engine/dml_with_input.go:35]:
```suggestion
	DML Primitive
```

File: go/vt/vtgate/planbuilder/operator_transformers.go [hunk 2]
```diff
@@ -685,66 +685,61 @@ func buildUpdateLogicalPlan(
...
-		Vindexes:          upd.VTable.ColumnVindexes,
-		OwnedVindexQuery:  upd.OwnedVindexQuery,
-		RoutingParameters: rp,
+	var vindexes []*vindexes.ColumnVindex
+	vQuery := ""
+	if len(upd.ChangedVindexValues) > 0 {
+		vQuery = sqlparser.String(upd.OwnedVindexQuery)
+		vindexes = upd.Target.VTable.ColumnVindexes
+		if upd.OwnedVindexQuery.Limit != nil && len(upd.OwnedVindexQuery.OrderBy) == 0 {
+			return nil, vterrors.VT12001("Vindex update should have ORDER BY clause when using LIMIT")
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operator_transformers.go:694]: is this an unsupported feature or a wrong input from the user?

Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/operator_transformers.go:694]: unsupported currently.

Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/operator_transformers.go:694]: as the output is non-deterministic.

Comment (frouioui) [reply go/vt/vtgate/planbuilder/operator_transformers.go:694]: got it

**Review by frouioui — APPROVED** [review]
this is a good change, thank you ! lgtm

File: go/vt/vtgate/planbuilder/operators/delete.go [hunk 3]
```diff
@@ -213,12 +216,41 @@ func createDeleteOperator(ctx *plancontext.PlanningContext, del *sqlparser.Delet
...
+func generateOwnedVindexQuery(tblExpr sqlparser.TableExpr, del *sqlparser.Delete, table TargetTable, ksidCols []sqlparser.IdentifierCI) *sqlparser.Select {
+	var selExprs sqlparser.SelectExprs
+	for _, col := range ksidCols {
+		colName := makeColName(col, table, sqlparser.MultiTable(del.TableExprs))
+		selExprs = append(selExprs, sqlparser.NewAliasedExpr(colName, ""))
+	}
+	for _, cv := range table.VTable.Owned {
+		for _, col := range cv.Columns {
+			colName := makeColName(col, table, sqlparser.MultiTable(del.TableExprs))
+			selExprs = append(selExprs, sqlparser.NewAliasedExpr(colName, ""))
```
Comment (systay) [inline go/vt/vtgate/planbuilder/operators/delete.go:228]: nitpick: aeWrap

File: go/vt/vtgate/planbuilder/operators/update.go [hunk 4]
```diff
@@ -150,66 +166,104 @@ func createUpdateOperator(ctx *plancontext.PlanningContext, updStmt *sqlparser.U
 		}
 	}
 
-	vp, cvv, ovq, subQueriesArgOnChangedVindex := getUpdateVindexInformation(ctx, updStmt, vindexTable, qt.ID, assignments)
+	target := updStmt.TableExprs[0]
+	atbl, ok := target.(*sqlparser.AliasedTableExpr)
```
Comment (systay) [inline go/vt/vtgate/planbuilder/operators/update.go:170]: shouldn't we not also check the size of `updStmt.TableExprs`? You can have multiple AliasedTableExpr next to each other by just coma separating the table names, I belive.

Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/operators/update.go:170]: currently, it is unsupported by the semantic analyzer.

File: go/vt/vtgate/planbuilder/testdata/foreignkey_cases.json [hunk 5]
```diff
@@ -3463,7 +3463,7 @@
               "Sharded": false
             },
             "FieldQuery": "select col14 from u_tbl1 where 1 != 1",
-            "Query": "select col14 from u_tbl1 where x = 2 and y = 4 lock in share mode",
+            "Query": "select /*+ SET_VAR(foreign_key_checks=OFF) */ col14 from u_tbl1 where x = 2 and y = 4 lock in share mode",
```
Comment (systay) [inline go/vt/vtgate/planbuilder/testdata/foreignkey_cases.json:3466]: what does this SET_VAR do here? `foreign_key_checks` isn't even applicable to `select`, right?

File: go/vt/vtgate/semantics/analyzer.go [hunk 6]
```diff
@@ -328,8 +328,8 @@ func (a *analyzer) noteQuerySignature(node sqlparser.SQLNode) {
 		}
 	case sqlparser.AggrFunc:
 		a.sig.Aggregation = true
-	case *sqlparser.Delete:
-		a.sig.Delete = true
+	case *sqlparser.Delete, *sqlparser.Update:
```
Comment (systay) [inline go/vt/vtgate/semantics/analyzer.go:331]: nitpick: might as well add the other DML as well - `INSERT`

**Review by systay — APPROVED** [review]

File: go/vt/vtgate/planbuilder/testdata/foreignkey_cases.json (hunk 5 above)
Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/testdata/foreignkey_cases.json:3466]: It is a side effect of having a `LockAndComment` operator. Should not cause any issue with it.

**Pull request merged**